beancount.el

```diff
@@ -527,8 +536,12 @@ will allow to align all numbers."
   (save-excursion
     (beginning-of-line)
     (cond
-     ;; Only timestamped directives start with a digit.
-     ((looking-at-p "[0-9]") 0)
+     ;; Lines which begin with a timestamp or a non-timestamped
+     ;; directive get indented to the beginning of the line.
+     ;; Consider: use "\\s-*[0-9]" instead of full timestamp regexp
+     ((or (looking-at-p beancount-timestamp-indent-regexp)
+          (looking-at-p beancount-directive-indent-regexp))
+      0)
```

Comment:
> As a general comment: beancount-mode assumes an opinionated style for the ledger. In this style directives are separated by empty lines and makes it possible to provide functionality without having to implement a complete Beancount syntax lexer and grammar in regular expressions.

I think that's a fair approach, but IMO if being less opinionated on what beancount-mode properly handles is not too onerous of a development effort and does not significantly adversely affect performance, then what's the downside? 

Personally, my ideal world would have beancount-mode able to behave much like `bean-format`, fixing all indentation where the intention is unambiguous. I'd also like an option to normalize all intraline whitespace (i.e. squash all consecutive spaces down to a single space except for when used for alignment), but I haven't looked deeply into this yet.

In general my philosophy with code that processes (potentially loosely) formatted data is to be lenient (within reason) with what you accept as input and strict with what you output.

I have been separating transactions by empty lines, which IMO looks the best and should probably be the preferred style, but I think other directives are better when tightly grouped together. That being said, this change is more-or-less unrelated to that (that is more relevant to `beancount-find-directive-extents` and friends).

> Can you please provide an example where this change improves editing?

The primary driver of this change was to explicitly check for non-timestamped directives to have 0 indentation.

> The choice of using the minimum possible context to decide the indentation level (ie only the first character in a date) is deliberate: it allows to indent lines that are only partially written. Why are you changing this?

The change from simply checking if the first character is a digit to checking if the line (after whitespace) began with a complete date was primarily to have "parallel structure" between the check for timestamped and non-timestamped directives for indentation purposes.

That being said, my change does not change the actual behavior for attempting to indent a line beginning with a partially completed date, as such lines would fall through to the default "0 indentation" case. And as you can see by my comment, I was not 100% convinced that "single-digit to full date" change was for the best, so I'm not going to put up a fight to keep the full date check (although I do think it's conceptually "cleaner", for lack of a better word).

I do, however, stand by the change to allow for arbitrary whitespace in front of the date/digit here, as this is a function whose purpose is to compute the appropriate indentation level of a line, and I do not think it makes sense to enforce that a line already have the appropriate indentation level when determining what its indentation level should be.

Reply:
What the mode is trying to indent should be valid Beancount syntax. An indented date in front of a directive is not valid Beancount syntax.

Reply:
I think that I fundamentally disagree with you here.

A function whose purpose is to properly indent a line but bases how to indent that line on how the line is already indented seems completely flawed to me.

While I'm not a *huge* fan of syntactically significant whitespace (probably because my primary language is C), I can understand it from a language spec perspective--trying to keep all files similarly formatted and legible is reasonable (although I think that doesn't matter as much with personal financial ledgers which are generally only viewed by one person or perhaps a small handful of people at most).

In the case of beancount, it's worth noting that the syntactically significant whitespace is completely unnecessary for unambiguously parsing a journal. I'm by no means suggesting that the beancount syntax should be changed to ignore leading whitespace, but I do want to point out the difference between beancount and, say, python, where changing the indentation of a line fundamentally changes that line's meaning.

For an editor whose goal is to try to make editing a language with syntactically significant whitespace as easy as possible, IMO there are essentially two reasonable choices for TAB/indentation behavior:

1. TAB toggles through indentation levels, possibly with separate commands to specifically shift a line (or region) up or down by an indentation level. This is the behavior of `python-mode`, and IMO should only be the behavior for languages where it is impossible to unambiguously determine what the "proper" indentation level of a given line is, i.e. if changing the indentation level completely changes a statement's meaning.

2. TAB indents the line (or region) as appropriate, regardless of the current indentation level. This is what `beancount-mode` currently does, with the exception of the case where a line happens to begin with 0 indentation. That distinction seems arbitrary to me, and IMO only serves to make editing more difficult.

The only reason the current behavior is even possible is because the beancount syntax only distinguishes between lines with leading whitespace and lines without leading whitespace, not caring about how much whitespace there is when whitespace is present. If there were multiple levels of syntactically significant indentation, then there would be no way for a given line to be automatically indented, as there wouldn't be a single correct indentation value for said line.


I'd also like to point out that despite your assertion that "what the mode is trying to indent should be valid Beancount syntax", the current indentation computation does not actually follow through with this in its behavior.

Consider the following snippet:

```
2021-01-01 * "Example"
Assets:Checking -1.00 USD
Expenses:Food 1.00 USD
```

This is clearly not valid Beancount syntax, as the postings are not indented. Despite this, trying to indent the "transaction" works as one would expect, changing it to valid Beancount syntax. (I use quotes around "transaction" here because, again, it is not valid syntax, and as far as Beancount is concerned it is *not* a transaction--it's not anything other than an error...).


Now I'd like you to consider to workflows, both of which I would consider reasonable yet are (at best) impeded by `beancount-mode`'s behavior:

1. Consider a brokerage account that holds both cash and a few different securities. As is often recommended, the user wants to have the one real account split into multiple sub-accounts by commodity type. It seems reasonable to me to want their `open` directives grouped together, without newlines separating them (and this is how Beancount's example journals are laid out):

   ```
   2021-01-01 open Assets:Brokerage:Cash               USD
   2021-01-01 open Assets:Brokerage:ABC                ABC
   2021-01-01 open Assets:Brokerage:XYZ                XYZ
   ```

   When inputting this in `beancount-mode` with `electric-indent-mode` enables (as is the default), one would type the first directive, then hit RET. Now the point is on the next line and indented. The user now types the second directive and hits RET. The second directive is still indented, despite this being invalid Beancount syntax. In order to get that second directive to be properly indented (i.e. to 0 indentation), the user would need to manually delete the indentation, either before or after inputting the directive itself. TAB doesn't help, because `beancount-mode` assumes that the correct indentation level is non-zero, despite the fact that it unambiguously is not. `beancount-mode` is actively working against the user, making it more difficult to edit their journal. The same process takes place for the next directive.

2. Let's say the user has `electric-indent-mode` turned off, either because they simply don't like its behavior in the general case or to avoid `beancount-mode` from working against them as in the previous example. The workflow that they want to use is to input a complete transaction and then indent it all at once, so they first type

   ```
   2021-01-01 * "Example"
   Assets:Checking  -1.00 USD
   posting_metadata: foo
   Expenses:Food 1.00 USD
   posting_metadata: bar
   ```

   Then they try to indent the transaction either with `M-q` or highlighting the region and hitting TAB or `C-M-\`. If my two-level metadata indentation changes were accepted without allowing metadata to be at zero-indentation when computing what its proper indentation level should be, the result would be the same as it is currently:

   ```
   2021-01-01 * "Example"
     Assets:Checking                           -1.00 USD
     posting_metadata: foo
     Expenses:Food                              1.00 USD
     posting_metadata: bar
   ```
   since the metadata would not match against the metadata regexp, and would then fall through to the "is the previous line indented" condition. Clearly this is not an improvement over the existing behavior.


All of that is to say that I feel quite strongly that `beancount-compute-indentation` should completely ignore the current line's existing implementation when determining what its proper indentation level should be.

That got much longer-winded than I initially intended, so I apologize for the wall of text.